// typeorm-a0744e6fec040634/commands/InitCommand.js
"use strict";var d=Object.defineProperty;var i=(c,e)=>d(c,"name",{value:e,configurable:!0});Object.defineProperty(exports,"__esModule",{value:!0}),exports.InitCommand=void 0;const tslib_1=require("tslib"),ansis_1=tslib_1.__importDefault(require("ansis")),node_path_1=tslib_1.__importDefault(require("node:path")),error_1=require("../error"),PlatformTools_1=require("../platform/PlatformTools"),CommandUtils_1=require("./CommandUtils");class InitCommand{static{i(this,"InitCommand")}constructor(){this.command="init",this.describe="Generates initial TypeORM project structure. If name specified then creates files inside directory called as name. If its not specified then creates files inside current directory."}builder(e){return e.option("n",{alias:"name",describe:"Name of the project directory."}).option("db",{alias:"database",describe:"Database type you'll use in your project.",choices:["postgres","mysql","mariadb","better-sqlite3","mssql","oracle","mongodb","cockroachdb","spanner"]}).option("express",{describe:"Indicates if express server sample code should be included in the project. False by default."}).option("docker",{describe:"Set to true if docker-compose must be generated as well. False by default."}).option("ms",{alias:"module",choices:["commonjs","esm"],default:"commonjs",describe:"Module system to use for project, expected values are commonjs or esm."})}async handler(e){try{const t=e.database??"postgres",r=e.express!==void 0,n=e.docker!==void 0,s=process.cwd()+(e.name?"/"+e.name:""),o=e.name?node_path_1.default.basename(e.name):void 0,a=e.ms==="esm";await CommandUtils_1.CommandUtils.createFile(s+"/package.json",InitCommand.getPackageJsonTemplate(o,a),!1),n&&await CommandUtils_1.CommandUtils.createFile(s+"/docker-compose.yml",InitCommand.getDockerComposeTemplate(t),!1),await CommandUtils_1.CommandUtils.createFile(s+"/.gitignore",InitCommand.getGitIgnoreFile()),await CommandUtils_1.CommandUtils.createFile(s+"/README.md",InitCommand.getReadmeTemplate({docker:n}),!1),await CommandUtils_1.CommandUtils.createFile(s+"/tsconfig.json",InitCommand.getTsConfigTemplate(a)),await CommandUtils_1.CommandUtils.createFile(s+"/src/entities/User.ts",InitCommand.getUserEntityTemplate(t)),await CommandUtils_1.CommandUtils.createFile(s+"/src/data-source.ts",InitCommand.getAppDataSourceTemplate(a,t)),await CommandUtils_1.CommandUtils.createFile(s+"/src/index.ts",InitCommand.getAppIndexTemplate(r,a)),await CommandUtils_1.CommandUtils.createDirectories(s+"/src/migrations"),r&&(await CommandUtils_1.CommandUtils.createFile(s+"/src/routes.ts",InitCommand.getRoutesTemplate(a)),await CommandUtils_1.CommandUtils.createFile(s+"/src/controllers/UserController.ts",InitCommand.getControllerTemplate(a)));const p=await CommandUtils_1.CommandUtils.readFile(s+"/package.json");await CommandUtils_1.CommandUtils.createFile(s+"/package.json",await InitCommand.appendPackageJson(p,t,r,a)),e.name?console.log(ansis_1.default.green`Project created inside ${ansis_1.default.blue(s)} directory.`):console.log(ansis_1.default.green`Project created inside current directory.`),console.log(ansis_1.default.green`Please verify the package.json file and install dependencies using your preferred package manager.`),console.log(ansis_1.default.green`Done! Start playing with a new project!`)}catch(t){PlatformTools_1.PlatformTools.logCmdErr("Error during project initialization:",t),process.exit(1)}}static getAppDataSourceTemplate(e,t){let r;switch(t){case"mysql":r=['type: "mysql"','host: "localhost"',"port: 3306",'username: "root"','password: "password"','database: "typeorm"'];break;case"mariadb":r=['type: "mariadb"','host: "localhost"',"port: 3306",'username: "root"','password: "password"','database: "typeorm"'];break;case"better-sqlite3":r=['type: "better-sqlite3"','database: "database.sqlite"'];break;case"postgres":r=['type: "postgres"','host: "localhost"',"port: 5432",'username: "root"','password: "password"','database: "typeorm"'];break;case"cockroachdb":r=['type: "cockroachdb"','host: "localhost"',"port: 26257",'username: "root"','password: ""','database: "defaultdb"'];break;case"mssql":r=['type: "mssql"','host: "localhost"','username: "sa"','password: "Admin12345"','database: "tempdb"'];break;case"oracle":r=['type: "oracle"','host: "localhost"','username: "system"','password: "oracle"',"port: 1521",'sid: "xe.oracle.docker"'];break;case"mongodb":r=['type: "mongodb"','database: "test"'];break;case"spanner":r=['type: "spanner"','projectId: "test"','instanceId: "test"','databaseId: "test"'];break;default:throw new Error(`Unknown database "${t}"`)}return`import "reflect-metadata"
import { DataSource } from "typeorm"
import { User } from "./entities/User${e?".js":""}"

export const AppDataSource = new DataSource({
${r.map(n=>`    ${n},`).join(`
`)}
    synchronize: true,
    logging: false,
    entities: [User],
    migrations: [],
    subscribers: [],
})
`}static getTsConfigTemplate(e){return JSON.stringify(e?{compilerOptions:{lib:["es2023"],target:"es2022",module:"es2022",moduleResolution:"node",allowSyntheticDefaultImports:!0,outDir:"./build",emitDecoratorMetadata:!0,experimentalDecorators:!0,sourceMap:!0}}:{compilerOptions:{lib:["es2023"],target:"es2022",module:"commonjs",moduleResolution:"node",outDir:"./build",emitDecoratorMetadata:!0,experimentalDecorators:!0,sourceMap:!0}},void 0,3)}static getGitIgnoreFile(){return`.idea/
.vscode/
node_modules/
build/
tmp/
temp/`}static getUserEntityTemplate(e){return`${e==="mongodb"?`import { Entity, ObjectIdColumn, Column } from "typeorm"
import { ObjectId } from "mongodb"`:'import { Entity, PrimaryGeneratedColumn, Column } from "typeorm"'}

@Entity()
export class User {

    ${e==="mongodb"?"@ObjectIdColumn()":"@PrimaryGeneratedColumn()"}
    id: ${e==="mongodb"?"ObjectId":"number"}

    @Column()
    firstName: string

    @Column()
    lastName: string

    @Column()
    age: number

}
`}static getRoutesTemplate(e){return`import { UserController } from "./controllers/UserController${e?".js":""}"

export const Routes = [{
    method: "get",
    route: "/users",
    controller: UserController,
    action: "all"
}, {
    method: "get",
    route: "/users/:id",
    controller: UserController,
    action: "one"
}, {
    method: "post",
    route: "/users",
    controller: UserController,
    action: "save"
}, {
    method: "delete",
    route: "/users/:id",
    controller: UserController,
    action: "remove"
}]`}static getControllerTemplate(e){return`import { AppDataSource } from "../data-source${e?".js":""}"
import { NextFunction, Request, Response } from "express"
import { User } from "../entities/User${e?".js":""}"

export class UserController {

    private userRepository = AppDataSource.getRepository(User)

    async all(request: Request, response: Response, next: NextFunction) {
        return this.userRepository.find()
    }

    async one(request: Request, response: Response, next: NextFunction) {
        const id = parseInt(request.params.id)


        const user = await this.userRepository.findOne({
            where: { id }
        })

        if (!user) {
            return "unregistered user"
        }
        return user
    }

    async save(request: Request, response: Response, next: NextFunction) {
        const { firstName, lastName, age } = request.body;

        const user = Object.assign(new User(), {
            firstName,
            lastName,
            age
        })

        return this.userRepository.save(user)
    }

    async remove(request: Request, response: Response, next: NextFunction) {
        const id = parseInt(request.params.id)

        let userToRemove = await this.userRepository.findOneBy({ id })

        if (!userToRemove) {
            return "this user not exist"
        }

        await this.userRepository.remove(userToRemove)

        return "user has been removed"
    }

}`}static getAppIndexTemplate(e,t){return e?`import ${t?"":"* as "}express from "express"
import ${t?"":"* as "}bodyParser from "body-parser"
import { Request, Response } from "express"
import { AppDataSource } from "./data-source${t?".js":""}"
import { Routes } from "./routes${t?".js":""}"
import { User } from "./entities/User${t?".js":""}"

AppDataSource.initialize().then(async () => {

    // create express app
    const app = express()
    app.use(bodyParser.json())

    // register express routes from defined application routes
    Routes.forEach(route => {
        (app as any)[route.method](route.route, (req: Request, res: Response, next: Function) => {
            const result = (new (route.controller as any))[route.action](req, res, next)
            if (result instanceof Promise) {
                result.then(result => result !== null && result !== undefined ? res.send(result) : undefined)

            } else if (result !== null && result !== undefined) {
                res.json(result)
            }
        })
    })

    // setup express app here
    // ...

    // start express server
    app.listen(3000)

    // insert new users for test
    await AppDataSource.manager.save(
        AppDataSource.manager.create(User, {
            firstName: "Timber",
            lastName: "Saw",
            age: 27
        })
    )

    await AppDataSource.manager.save(
        AppDataSource.manager.create(User, {
            firstName: "Phantom",
            lastName: "Assassin",
            age: 24
        })
    )

    console.log("Express server has started on port 3000. Open http://localhost:3000/users to see results")

}).catch(error => console.log(error))
`:`import { AppDataSource } from "./data-source${t?".js":""}"
import { User } from "./entities/User${t?".js":""}"

AppDataSource.initialize().then(async () => {

    console.log("Inserting a new user into the database...")
    const user = new User()
    user.firstName = "Timber"
    user.lastName = "Saw"
    user.age = 25
    await AppDataSource.manager.save(user)
    console.log("Saved a new user with id: " + user.id)

    console.log("Loading users from the database...")
    const users = await AppDataSource.manager.find(User)
    console.log("Loaded users: ", users)

    console.log("Here you can setup and run express / fastify / any other framework.")

}).catch(error => console.log(error))
`}static getPackageJsonTemplate(e,t){return JSON.stringify({name:e??"typeorm-sample",version:"0.0.1",description:"Awesome project developed with TypeORM.",type:t?"module":"commonjs",devDependencies:{},dependencies:{},scripts:{}},void 0,3)}static getDockerComposeTemplate(e){switch(e){case"mysql":return`services:

  mysql:
    image: "mysql:9.2.0"
    ports:
      - "3306:3306"
    environment:
      MYSQL_ROOT_PASSWORD: "password"
      MYSQL_DATABASE: "typeorm"

`;case"mariadb":return`services:

  mariadb:
    image: "mariadb:11.7.2"
    ports:
      - "3306:3306"
    environment:
      MYSQL_ROOT_PASSWORD: "password"
      MYSQL_DATABASE: "typeorm"

`;case"postgres":return`services:

  postgres:
    image: "postgres:17.2"
    ports:
      - "5432:5432"
    environment:
      POSTGRES_USER: "root"
      POSTGRES_PASSWORD: "password"
      POSTGRES_DB: "typeorm"

`;case"cockroachdb":return`services:

  cockroachdb:
    image: "cockroachdb/cockroach:v25.1.2"
    command: start --insecure
    ports:
      - "26257:26257"

`;case"better-sqlite3":throw new error_1.TypeORMError("SQLite does not require docker");case"oracle":throw new error_1.TypeORMError("You cannot initialize a project with docker for Oracle driver yet.");case"mssql":return`services:

  mssql:
    image: "mcr.microsoft.com/mssql/server:2022-latest"
    ports:
      - "1433:1433"
    environment:
      SA_PASSWORD: "Admin12345"
      ACCEPT_EULA: "Y"
      MSSQL_PID: "Express"

`;case"mongodb":return`services:

  mongodb:
    image: "mongo:8"
    container_name: "typeorm-mongodb"
    ports:
      - "27017:27017"

`;case"spanner":return`services:

  spanner:
    image: gcr.io/cloud-spanner-emulator/emulator:1.5.30
    ports:
      - "9010:9010"
      - "9020:9020"

`}return""}static getReadmeTemplate(e){let t=`# Awesome Project Build with TypeORM

Steps to run this project:

1. Run \`npm i\` command
`;return e.docker?t+="2. Run `docker-compose up` command\n":t+="2. Setup database settings inside `data-source.ts` file\n",t+="3. Run `npm start` command\n",t}static async appendPackageJson(e,t,r,n){const s=JSON.parse(e),o=JSON.parse(await CommandUtils_1.CommandUtils.readFile(node_path_1.default.resolve(__dirname,"..","package.json")));switch(s.devDependencies??={},s.devDependencies={"@types/node":o.devDependencies["@types/node"],"ts-node":o.devDependencies["ts-node"],typescript:o.devDependencies.typescript,...s.devDependencies},s.dependencies??={},s.dependencies={...s.dependencies,"reflect-metadata":o.dependencies["reflect-metadata"],typeorm:o.version},t){case"mysql":case"mariadb":s.dependencies.mysql2=o.devDependencies.mysql2;break;case"postgres":case"cockroachdb":s.dependencies.pg=o.devDependencies.pg;break;case"better-sqlite3":s.dependencies["better-sqlite3"]=o.devDependencies["better-sqlite3"];break;case"oracle":s.dependencies.oracledb=o.devDependencies.oracledb;break;case"mssql":s.dependencies.mssql=o.devDependencies.mssql;break;case"mongodb":s.dependencies.mongodb=o.devDependencies.mongodb;break;case"spanner":s.dependencies["@google-cloud/spanner"]=o.devDependencies["@google-cloud/spanner"];break}return r&&(s.dependencies.express="^4.21.2",s.dependencies["body-parser"]="^1.20.3"),s.scripts??={},n?Object.assign(s.scripts,{start:"node --loader ts-node/esm src/index.ts",typeorm:"typeorm-ts-node-esm"}):Object.assign(s.scripts,{start:"ts-node src/index.ts",typeorm:"typeorm-ts-node-commonjs"}),JSON.stringify(s,void 0,3)}}exports.InitCommand=InitCommand;
