// next/dist/lib/verify-root-layout.js
"use strict";Object.defineProperty(exports,"__esModule",{value:!0}),Object.defineProperty(exports,"verifyRootLayout",{enumerable:!0,get:function(){return verifyRootLayout}});const _path=_interop_require_default(require("node:path")),_log=_interop_require_wildcard(require("../build/output/log")),_fs=require("node:fs"),_picocolors=require("./picocolors"),_constants=require("./constants");function _interop_require_default(e){return e&&e.__esModule?e:{default:e}}function _getRequireWildcardCache(e){if(typeof WeakMap!="function")return null;var n=new WeakMap,o=new WeakMap;return(_getRequireWildcardCache=function(r){return r?o:n})(e)}function _interop_require_wildcard(e,n){if(!n&&e&&e.__esModule)return e;if(e===null||typeof e!="object"&&typeof e!="function")return{default:e};var o=_getRequireWildcardCache(n);if(o&&o.has(e))return o.get(e);var r={__proto__:null},s=Object.defineProperty&&Object.getOwnPropertyDescriptor;for(var t in e)if(t!=="default"&&Object.prototype.hasOwnProperty.call(e,t)){var i=s?Object.getOwnPropertyDescriptor(e,t):null;i&&(i.get||i.set)?Object.defineProperty(r,t,i):r[t]=e[t]}return r.default=e,o&&o.set(e,r),r}const globOrig=require("next/dist/compiled/glob"),glob=(e,n)=>new Promise((o,r)=>{globOrig(n,{cwd:e},(s,t)=>{if(s)return r(s);o(t)})});function getRootLayout(e){return e?`export const metadata = {
  title: 'Next.js',
  description: 'Generated by Next.js',
}

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html lang="en">
      <body>{children}</body>
    </html>
  )
}
`:`export const metadata = {
  title: 'Next.js',
  description: 'Generated by Next.js',
}

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body>{children}</body>
    </html>
  )
}
`}async function verifyRootLayout({dir:e,appDir:n,tsconfigPath:o,pagePath:r,pageExtensions:s}){let t;try{const i=await glob(n,`**/layout.{${s.join(",")}}`),d=r.startsWith(`${_constants.APP_DIR_ALIAS}/`),p=r.replace(`${_constants.APP_DIR_ALIAS}/`,""),c=p.split("/");let u;if(d)if(i.length===0){const a=c[0];u=a.startsWith("(")?a:""}else{c.pop();let a=[];for(const f of c)if(a.push(f),!i.some(l=>l.startsWith(a.join("/")))){u=a.join("/");break}}else u="";if(typeof u=="string"){const a=o||"tsconfig.json",f=_path.default.join(e,a),l=await _fs.promises.access(f).then(()=>!0,()=>!1);return t=_path.default.join(n,u,`layout.${l?"tsx":"js"}`),await _fs.promises.writeFile(t,getRootLayout(l)),_log.warn(`Your page ${(0,_picocolors.bold)(`app/${p}`)} did not have a root layout. We created ${(0,_picocolors.bold)(`app${t.replace(n,"")}`)} for you.`),[!0,t]}}catch(i){console.error(i)}return[!1,t]}
