// next/dist/server/lib/router-utils/cache-life-type-utils.js
"use strict";var o=Object.defineProperty;var a=(e,i)=>o(e,"name",{value:i,configurable:!0});Object.defineProperty(exports,"__esModule",{value:!0});function _export(e,i){for(var r in i)Object.defineProperty(e,r,{enumerable:!0,get:i[r]})}a(_export,"_export"),_export(exports,{generateCacheLifeTypes:a(function(){return generateCacheLifeTypes},"generateCacheLifeTypes"),writeCacheLifeTypes:a(function(){return writeCacheLifeTypes},"writeCacheLifeTypes")});const _fs=_interop_require_default(require("node:fs")),_path=_interop_require_default(require("node:path"));function _interop_require_default(e){return e&&e.__esModule?e:{default:e}}a(_interop_require_default,"_interop_require_default");const MINUTE_IN_SECONDS=60,HOUR_IN_SECONDS=MINUTE_IN_SECONDS*60,DAY_IN_SECONDS=HOUR_IN_SECONDS*24,WEEK_IN_SECONDS=DAY_IN_SECONDS*7,MONTH_30_DAYS_IN_SECONDS=DAY_IN_SECONDS*30;function formatTimespan(e){if(e>0){if(e===MONTH_30_DAYS_IN_SECONDS)return"1 month";if(e===WEEK_IN_SECONDS)return"1 week";if(e===DAY_IN_SECONDS)return"1 day";if(e===HOUR_IN_SECONDS)return"1 hour";if(e===MINUTE_IN_SECONDS)return"1 minute";if(e%MONTH_30_DAYS_IN_SECONDS===0)return e/MONTH_30_DAYS_IN_SECONDS+" months";if(e%18144e3===0)return e/18144e3+" months";if(e%WEEK_IN_SECONDS===0)return e/WEEK_IN_SECONDS+" weeks";if(e%DAY_IN_SECONDS===0)return e/DAY_IN_SECONDS+" days";if(e%HOUR_IN_SECONDS===0)return e/HOUR_IN_SECONDS+" hours";if(e%MINUTE_IN_SECONDS===0)return e/MINUTE_IN_SECONDS+" minutes"}return e+" seconds"}a(formatTimespan,"formatTimespan");function formatTimespanWithSeconds(e){if(e===void 0)return"default";if(e>=4294967294)return"never";const i=e+" seconds",r=formatTimespan(e);return r===i?i:i+" ("+r+")"}a(formatTimespanWithSeconds,"formatTimespanWithSeconds");function generateCacheLifeTypes(e){let i="";const r=Object.keys(e);for(let f=0;f<r.length;f++){const s=r[f],t=e[s];if(typeof t!="object"||t===null)continue;let n="";t.stale===void 0?n+=`
     * This cache may be stale on clients for the default stale time of the scope before checking with the server.`:t.stale>=4294967294?n+=`
     * This cache may be stale on clients indefinitely before checking with the server.`:n+=`
     * This cache may be stale on clients for ${formatTimespan(t.stale)} before checking with the server.`,t.revalidate!==void 0&&t.expire!==void 0&&t.revalidate>=t.expire?n+=`
     * This cache will expire after ${formatTimespan(t.expire)}. The next request will recompute it.`:(t.revalidate===void 0?n+=`
     * It will inherit the default revalidate time of its scope since it does not define its own.`:t.revalidate>=4294967294||(n+=`
     * If the server receives a new request after ${formatTimespan(t.revalidate)}, start revalidating new values in the background.`),t.expire===void 0?n+=`
     * It will inherit the default expiration time of its scope since it does not define its own.`:t.expire>=4294967294?n+=`
     * It lives for the maximum age of the server cache. If this entry has no traffic for a while, it may serve an old value the next request.`:n+=`
     * If this entry has no traffic for ${formatTimespan(t.expire)} it will expire. The next request will recompute it.`),i+=`
    /**
     * Cache this \`"use cache"\` for a timespan defined by the \`${JSON.stringify(s)}\` profile.
     * \`\`\`
     *   stale:      ${formatTimespanWithSeconds(t.stale)}
     *   revalidate: ${formatTimespanWithSeconds(t.revalidate)}
     *   expire:     ${formatTimespanWithSeconds(t.expire)}
     * \`\`\`
     * ${n}
     */
    export function cacheLife(profile: ${JSON.stringify(s)}): void
    `}return i+=`
    /**
     * Cache this \`"use cache"\` using a custom timespan.
     * \`\`\`
     *   stale: ... // seconds
     *   revalidate: ... // seconds
     *   expire: ... // seconds
     * \`\`\`
     *
     * This is similar to Cache-Control: max-age=\`stale\`,s-max-age=\`revalidate\`,stale-while-revalidate=\`expire-revalidate\`
     *
     * If a value is left out, the lowest of other cacheLife() calls or the default, is used instead.
     */
    export function cacheLife(profile: {
      /**
       * This cache may be stale on clients for ... seconds before checking with the server.
       */
      stale?: number,
      /**
       * If the server receives a new request after ... seconds, start revalidating new values in the background.
       */
      revalidate?: number,
      /**
       * If this entry has no traffic for ... seconds it will expire. The next request will recompute it.
       */
      expire?: number
    }): void
  `,`// Type definitions for Next.js cacheLife configs

declare module 'next/cache' {
  export { unstable_cache } from 'next/dist/server/web/spec-extension/unstable-cache'
  export {
    updateTag,
    revalidateTag,
    revalidatePath,
    refresh,
  } from 'next/dist/server/web/spec-extension/revalidate'
  export { unstable_noStore } from 'next/dist/server/web/spec-extension/unstable-no-store'

  ${i}

  import { cacheTag } from 'next/dist/server/use-cache/cache-tag'
  export { cacheTag }

  export const unstable_cacheTag: typeof cacheTag
  export const unstable_cacheLife: typeof cacheLife
}
`}a(generateCacheLifeTypes,"generateCacheLifeTypes");function writeCacheLifeTypes(e,i){if(!e||Object.keys(e).length===0)return;const r=_path.default.dirname(i);_fs.default.existsSync(r)||_fs.default.mkdirSync(r,{recursive:!0});const f=generateCacheLifeTypes(e);_fs.default.writeFileSync(i,f)}a(writeCacheLifeTypes,"writeCacheLifeTypes");
