// next/dist/server/lib/router-utils/typegen.js
"use strict";Object.defineProperty(exports,"__esModule",{value:!0});function _export(e,o){for(var t in o)Object.defineProperty(e,t,{enumerable:!0,get:o[t]})}_export(exports,{generateLinkTypesFile:function(){return generateLinkTypesFile},generateRouteTypesFile:function(){return generateRouteTypesFile},generateRouteTypesFileStrict:function(){return generateRouteTypesFileStrict},generateValidatorFile:function(){return generateValidatorFile},generateValidatorFileStrict:function(){return generateValidatorFileStrict}});const _isdynamic=require("../../../shared/lib/router/utils/is-dynamic");function generateRouteTypes(e){const o=Object.keys(e.appRoutes).sort(),t=Object.keys(e.pageRoutes).sort(),s=Object.keys(e.layoutRoutes).sort(),a=Object.keys(e.redirectRoutes).sort(),n=Object.keys(e.rewriteRoutes).sort();let r="";o.length>0?r+=`type AppRoutes = ${o.map(u=>JSON.stringify(u)).join(" | ")}
`:r+=`type AppRoutes = never
`;const l=Object.keys(e.appRouteHandlerRoutes).sort(),p=l.length>0;p&&(r+=`type AppRouteHandlerRoutes = ${l.map(u=>JSON.stringify(u)).join(" | ")}
`),t.length>0?r+=`type PageRoutes = ${t.map(u=>JSON.stringify(u)).join(" | ")}
`:r+=`type PageRoutes = never
`,s.length>0?r+=`type LayoutRoutes = ${s.map(u=>JSON.stringify(u)).join(" | ")}
`:r+=`type LayoutRoutes = never
`,a.length>0?r+=`type RedirectRoutes = ${a.map(u=>JSON.stringify(u)).join(" | ")}
`:r+=`type RedirectRoutes = never
`,n.length>0?r+=`type RewriteRoutes = ${n.map(u=>JSON.stringify(u)).join(" | ")}
`:r+=`type RewriteRoutes = never
`;const i=["AppRoutes","PageRoutes","LayoutRoutes","RedirectRoutes","RewriteRoutes"];return p&&i.push("AppRouteHandlerRoutes"),r+=`type Routes = ${i.join(" | ")}
`,r}function generateParamTypes(e){const o={...e.appRoutes,...e.appRouteHandlerRoutes,...e.pageRoutes,...e.layoutRoutes,...e.redirectRoutes,...e.rewriteRoutes};let t=`interface ParamMap {
`;const s=Object.entries(o).sort(([a],[n])=>a.localeCompare(n));for(const[a,n]of s){const{groups:r}=n;if(!(0,_isdynamic.isDynamicRoute)(a)||Object.keys(r??{}).length===0){t+=`  ${JSON.stringify(a)}: {}
`;continue}let l="{";for(const[p,i]of Object.entries(r)){const u=JSON.stringify(p);i.repeat?i.optional?l+=` ${u}?: string[];`:l+=` ${u}: string[];`:i.optional?l+=` ${u}?: string;`:l+=` ${u}: string;`}l+=" }",t+=`  ${JSON.stringify(a)}: ${l}
`}return t+=`}
`,t}function generateLayoutSlotMap(e){let o=`interface LayoutSlotMap {
`;const t=Object.entries(e.layoutRoutes).sort(([s],[a])=>s.localeCompare(a));for(const[s,a]of t)if("slots"in a){const n=a.slots.sort();n.length>0?o+=`  ${JSON.stringify(s)}: ${n.map(r=>JSON.stringify(r)).join(" | ")}
`:o+=`  ${JSON.stringify(s)}: never
`}else o+=`  ${JSON.stringify(s)}: never
`;return o+=`}
`,o}function formatRouteToRouteType(e){const o=(0,_isdynamic.isDynamicRoute)(e);return o&&(e=e.split("/").map(t=>t.startsWith("[")&&t.endsWith("]")?t.startsWith("[...")?"${CatchAllSlug<T>}":t.startsWith("[[...")&&t.endsWith("]]")?"${OptionalCatchAllSlug<T>}":"${SafeSlug<T>}":t).join("/")),{isDynamic:o,routeType:e}}function serializeRouteTypes(e){return e.sort().map(o=>`
    | \`${o}\``).join("")}function generateLinkTypesFile(e){const o=new Set([...Object.keys(e.appRoutes),...Object.keys(e.pageRoutes),...Object.keys(e.redirectRoutes),...Object.keys(e.rewriteRoutes),...Object.keys(e.appRouteHandlerRoutes),...Array.from(e.pageApiRoutes)]),t=[],s=[];for(const l of o){const{isDynamic:p,routeType:i}=formatRouteToRouteType(l);p?s.push(i):t.push(i)}const a=serializeRouteTypes(t),n=serializeRouteTypes(s);return`// This file is generated automatically by Next.js
// Do not edit this file manually

// Type definitions for Next.js routes

/**
 * Internal types used by the Next.js router and Link component.
 * These types are not meant to be used directly.
 * @internal
 */
declare namespace __next_route_internal_types__ {
  type SearchOrHash = \`?\${string}\` | \`#\${string}\`
  type WithProtocol = \`\${string}:\${string}\`

  type Suffix = '' | SearchOrHash

  type SafeSlug<S extends string> = S extends \`\${string}/\${string}\`
    ? never
    : S extends \`\${string}\${SearchOrHash}\`
    ? never
    : S extends ''
    ? never
    : S

  type CatchAllSlug<S extends string> = S extends \`\${string}\${SearchOrHash}\`
    ? never
    : S extends ''
    ? never
    : S

  type OptionalCatchAllSlug<S extends string> =
    S extends \`\${string}\${SearchOrHash}\` ? never : S

  type StaticRoutes = ${a||"never"}
  type DynamicRoutes<T extends string = string> = ${n||"never"}

  type RouteImpl<T> = ${(!a&&!n?"string & {}":"")||"\n    | StaticRoutes\n    | SearchOrHash\n    | WithProtocol\n    | `${StaticRoutes}${SearchOrHash}`\n    | (T extends `${DynamicRoutes<infer _>}${Suffix}` ? T : never)\n    "}
}

declare module 'next' {
  export { default } from 'next/types.js'
  export * from 'next/types.js'

  export type Route<T extends string = string> =
    __next_route_internal_types__.RouteImpl<T>
}

declare module 'next/link' {
  export { useLinkStatus } from 'next/dist/client/link.js'

  import type { LinkProps as OriginalLinkProps } from 'next/dist/client/link.js'
  import type { AnchorHTMLAttributes, DetailedHTMLProps } from 'react'
  import type { UrlObject } from 'url'

  type LinkRestProps = Omit<
    Omit<
      DetailedHTMLProps<
        AnchorHTMLAttributes<HTMLAnchorElement>,
        HTMLAnchorElement
      >,
      keyof OriginalLinkProps
    > &
      OriginalLinkProps,
    'href'
  >

  export type LinkProps<RouteInferType> = LinkRestProps & {
    /**
     * The path or URL to navigate to. This is the only required prop. It can also be an object.
     * @see https://nextjs.org/docs/api-reference/next/link
     */
    href: __next_route_internal_types__.RouteImpl<RouteInferType> | UrlObject
  }

  export default function Link<RouteType>(props: LinkProps<RouteType>): JSX.Element
}

declare module 'next/navigation' {
  export * from 'next/dist/client/components/navigation.js'

  import type { NavigateOptions, AppRouterInstance as OriginalAppRouterInstance } from 'next/dist/shared/lib/app-router-context.shared-runtime.js'
  import type { RedirectType } from 'next/dist/client/components/redirect-error.js'
  
  interface AppRouterInstance extends OriginalAppRouterInstance {
    /**
     * Navigate to the provided href.
     * Pushes a new history entry.
     */
    push<RouteType>(href: __next_route_internal_types__.RouteImpl<RouteType>, options?: NavigateOptions): void
    /**
     * Navigate to the provided href.
     * Replaces the current history entry.
     */
    replace<RouteType>(href: __next_route_internal_types__.RouteImpl<RouteType>, options?: NavigateOptions): void
    /**
     * Prefetch the provided href.
     */
    prefetch<RouteType>(href: __next_route_internal_types__.RouteImpl<RouteType>): void
  }

  export function useRouter(): AppRouterInstance;
  
  /**
   * This function allows you to redirect the user to another URL. It can be used in
   * [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components),
   * [Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers), and
   * [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations).
   *
   * - In a Server Component, this will insert a meta tag to redirect the user to the target page.
   * - In a Route Handler or Server Action, it will serve a 307/303 to the caller.
   * - In a Server Action, type defaults to 'push' and 'replace' elsewhere.
   *
   * Read more: [Next.js Docs: redirect](https://nextjs.org/docs/app/api-reference/functions/redirect)
   */
  export function redirect<RouteType>(
    /** The URL to redirect to */
    url: __next_route_internal_types__.RouteImpl<RouteType>,
    type?: RedirectType
  ): never;
  
  /**
   * This function allows you to redirect the user to another URL. It can be used in
   * [Server Components](https://nextjs.org/docs/app/building-your-application/rendering/server-components),
   * [Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers), and
   * [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations).
   *
   * - In a Server Component, this will insert a meta tag to redirect the user to the target page.
   * - In a Route Handler or Server Action, it will serve a 308/303 to the caller.
   *
   * Read more: [Next.js Docs: redirect](https://nextjs.org/docs/app/api-reference/functions/redirect)
   */
  export function permanentRedirect<RouteType>(
    /** The URL to redirect to */
    url: __next_route_internal_types__.RouteImpl<RouteType>,
    type?: RedirectType
  ): never;
}

declare module 'next/form' {
  import type { FormProps as OriginalFormProps } from 'next/dist/client/form.js'

  type FormRestProps = Omit<OriginalFormProps, 'action'>

  export type FormProps<RouteInferType> = {
    /**
     * \`action\` can be either a \`string\` or a function.
     * - If \`action\` is a string, it will be interpreted as a path or URL to navigate to when the form is submitted.
     *   The path will be prefetched when the form becomes visible.
     * - If \`action\` is a function, it will be called when the form is submitted. See the [React docs](https://react.dev/reference/react-dom/components/form#props) for more.
     */
    action: __next_route_internal_types__.RouteImpl<RouteInferType> | ((formData: FormData) => void)
  } & FormRestProps

  export default function Form<RouteType>(props: FormProps<RouteType>): JSX.Element
}
`}function generateValidatorFile(e){const o=(P,R,m)=>P.sort().filter(c=>c.endsWith(".ts")||c.endsWith(".tsx")).filter(c=>R!=="AppPageConfig"||c.endsWith("page.ts")||c.endsWith("page.tsx")).map(c=>{const x=c,y=m?.get(c),h=y&&(R==="AppPageConfig"||R==="LayoutConfig"||R==="RouteHandlerConfig")?`${R}<${JSON.stringify(y)}>`:R;return`// Validate ${c}
{
  type __IsExpected<Specific extends ${h}> = Specific
  const handler = {} as typeof import(${JSON.stringify(x.replace(/\.tsx?$/,".js"))})
  type __Check = __IsExpected<typeof handler>
  // @ts-ignore
  type __Unused = __Check
}`}).join(`

`),t=o(Array.from(e.appPagePaths).sort(),"AppPageConfig",e.filePathToRoute),s=o(Array.from(e.appRouteHandlers).sort(),"RouteHandlerConfig",e.filePathToRoute),a=o(Array.from(e.pagesRouterPagePaths).sort(),"PagesPageConfig"),n=o(Array.from(e.pageApiRoutes).sort(),"ApiRouteConfig"),r=o(Array.from(e.layoutPaths).sort(),"LayoutConfig",e.filePathToRoute),l=Object.keys(e.appRouteHandlerRoutes).length>0;let p="";t&&(p+=`type AppPageConfig<Route extends AppRoutes = AppRoutes> = {
  default: React.ComponentType<{ params: Promise<ParamMap[Route]> } & any> | ((props: { params: Promise<ParamMap[Route]> } & any) => React.ReactNode | Promise<React.ReactNode> | never | void | Promise<void>)
  generateStaticParams?: (props: { params: ParamMap[Route] }) => Promise<any[]> | any[]
  generateMetadata?: (
    props: { params: Promise<ParamMap[Route]> } & any,
    parent: ResolvingMetadata
  ) => Promise<any> | any
  generateViewport?: (
    props: { params: Promise<ParamMap[Route]> } & any,
    parent: ResolvingViewport
  ) => Promise<any> | any
  metadata?: any
  viewport?: any
}

`),a&&(p+=`type PagesPageConfig = {
  default: React.ComponentType<any> | ((props: any) => React.ReactNode | Promise<React.ReactNode> | never | void)
  getStaticProps?: (context: any) => Promise<any> | any
  getStaticPaths?: (context: any) => Promise<any> | any
  getServerSideProps?: (context: any) => Promise<any> | any
  getInitialProps?: (context: any) => Promise<any> | any
  /**
   * Segment configuration for legacy Pages Router pages.
   * Validated at build-time by parsePagesSegmentConfig.
   */
  config?: {
    maxDuration?: number
    runtime?: 'edge' | 'experimental-edge' | 'nodejs' | string // necessary unless config is exported as const
    regions?: string[]
  }
}

`),r&&(p+=`type LayoutConfig<Route extends LayoutRoutes = LayoutRoutes> = {
  default: React.ComponentType<LayoutProps<Route>> | ((props: LayoutProps<Route>) => React.ReactNode | Promise<React.ReactNode> | never | void | Promise<void>)
  generateStaticParams?: (props: { params: ParamMap[Route] }) => Promise<any[]> | any[]
  generateMetadata?: (
    props: { params: Promise<ParamMap[Route]> } & any,
    parent: ResolvingMetadata
  ) => Promise<any> | any
  generateViewport?: (
    props: { params: Promise<ParamMap[Route]> } & any,
    parent: ResolvingViewport
  ) => Promise<any> | any
  metadata?: any
  viewport?: any
}

`),s&&(p+=`type RouteHandlerConfig<Route extends AppRouteHandlerRoutes = AppRouteHandlerRoutes> = {
  GET?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  POST?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  PUT?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  PATCH?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  DELETE?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  HEAD?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  OPTIONS?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
}

`),n&&(p+=`type ApiRouteConfig = {
  default: (req: any, res: any) => ReturnType<NextApiHandler>
  config?: {
    api?: {
      bodyParser?: boolean | { sizeLimit?: number | string }
      responseLimit?: string | number | boolean
      externalResolver?: boolean
    }
    runtime?: 'edge' | 'experimental-edge' | 'nodejs' | string // necessary unless config is exported as const
    maxDuration?: number
  }
}

`);const i=[];t&&i.push("AppRoutes"),r&&i.push("LayoutRoutes"),(t||r||s)&&i.push("ParamMap"),l&&i.push("AppRouteHandlerRoutes");const u=i.length>0?`import type { ${i.join(", ")} } from "./routes.js"`:"",g=l?`import type { NextRequest } from 'next/server.js'
`:"",d=[];n&&d.push("NextApiHandler"),(t||r)&&d.push("ResolvingMetadata","ResolvingViewport");const f=d.length>0?`import type { ${d.join(", ")} } from "next/types.js"
`:"";return`// This file is generated automatically by Next.js
// Do not edit this file manually
// This file validates that all pages and layouts export the correct types

${u}
${f}${g}
${p}
${t}

${s}

${a}

${n}

${r}
`}function generateValidatorFileStrict(e){const o=(P,R,m)=>P.sort().filter(c=>c.endsWith(".ts")||c.endsWith(".tsx")).filter(c=>R!=="AppPageConfig"||c.endsWith("page.ts")||c.endsWith("page.tsx")).map(c=>{const x=c,y=m?.get(c),h=y&&(R==="AppPageConfig"||R==="LayoutConfig"||R==="RouteHandlerConfig")?`${R}<${JSON.stringify(y)}>`:R;return`// Validate ${c}
{
  const handler = {} as typeof import(${JSON.stringify(x.replace(/\.tsx?$/,".js"))})
  handler satisfies ${h}
}`}).join(`

`),t=o(Array.from(e.appPagePaths).sort(),"AppPageConfig",e.filePathToRoute),s=o(Array.from(e.appRouteHandlers).sort(),"RouteHandlerConfig",e.filePathToRoute),a=o(Array.from(e.pagesRouterPagePaths).sort(),"PagesPageConfig"),n=o(Array.from(e.pageApiRoutes).sort(),"ApiRouteConfig"),r=o(Array.from(e.layoutPaths).sort(),"LayoutConfig",e.filePathToRoute),l=Object.keys(e.appRouteHandlerRoutes).length>0;let p="";t&&(p+=`type AppPageConfig<Route extends AppRoutes = AppRoutes> = {
  default: React.JSXElementConstructor<PageProps<Route>>
  generateStaticParams?: (props: { params: ParamMap[Route] }) => Promise<any[]> | any[]
  generateMetadata?: (
    props: PageProps<Route>,
    parent: ResolvingMetadata
  ) => Promise<any> | any
  generateViewport?: (
    props: PageProps<Route>,
    parent: ResolvingViewport
  ) => Promise<any> | any
  metadata?: any
  viewport?: any
}

`),a&&(p+=`type PagesPageConfig = {
  default: React.ComponentType<any> | ((props: any) => React.ReactNode | Promise<React.ReactNode> | never | void)
  getStaticProps?: (context: any) => Promise<any> | any
  getStaticPaths?: (context: any) => Promise<any> | any
  getServerSideProps?: (context: any) => Promise<any> | any
  getInitialProps?: (context: any) => Promise<any> | any
  /**
   * Segment configuration for legacy Pages Router pages.
   * Validated at build-time by parsePagesSegmentConfig.
   */
  config?: {
    maxDuration?: number
    runtime?: 'edge' | 'experimental-edge' | 'nodejs' | string // necessary unless config is exported as const
    regions?: string[]
  }
}

`),r&&(p+=`type LayoutConfig<Route extends LayoutRoutes = LayoutRoutes> = {
  default: React.ComponentType<LayoutProps<Route>> | ((props: LayoutProps<Route>) => React.ReactNode | Promise<React.ReactNode> | never | void | Promise<void>)
  generateStaticParams?: (props: { params: ParamMap[Route] }) => Promise<any[]> | any[]
  generateMetadata?: (
    props: { params: Promise<ParamMap[Route]> } & any,
    parent: ResolvingMetadata
  ) => Promise<any> | any
  generateViewport?: (
    props: { params: Promise<ParamMap[Route]> } & any,
    parent: ResolvingViewport
  ) => Promise<any> | any
  metadata?: any
  viewport?: any
}

`),s&&(p+=`type RouteHandlerConfig<Route extends AppRouteHandlerRoutes = AppRouteHandlerRoutes> = {
  GET?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  POST?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  PUT?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  PATCH?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  DELETE?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  HEAD?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
  OPTIONS?: (request: NextRequest, context: { params: Promise<ParamMap[Route]> }) => Promise<Response | void> | Response | void
}

`),n&&(p+=`type ApiRouteConfig = {
  default: (req: any, res: any) => ReturnType<NextApiHandler>
  config?: {
    api?: {
      bodyParser?: boolean | { sizeLimit?: number | string }
      responseLimit?: string | number | boolean
      externalResolver?: boolean
    }
    runtime?: 'edge' | 'experimental-edge' | 'nodejs' | string // necessary unless config is exported as const
    maxDuration?: number
  }
}

`);const i=[];t&&i.push("AppRoutes"),r&&i.push("LayoutRoutes"),(t||r||s)&&i.push("ParamMap"),l&&i.push("AppRouteHandlerRoutes");const u=i.length>0?`import type { ${i.join(", ")} } from "./routes.js"`:"",g=l?`import type { NextRequest } from 'next/server.js'
`:"",d=[];n&&d.push("NextApiHandler"),(t||r)&&d.push("ResolvingMetadata","ResolvingViewport");const f=d.length>0?`import type { ${d.join(", ")} } from "next/types.js"
`:"";return`// This file is generated automatically by Next.js with experimental.strictRouteTypes
// Do not edit this file manually
// This file validates that all pages and layouts export the correct types

${u}
${f}${g}
${p}
${t}

${s}

${a}

${n}

${r}
`}function generateRouteTypesFile(e){const o=generateRouteTypes(e),t=generateParamTypes(e),s=generateLayoutSlotMap(e),a=Object.keys(e.appRouteHandlerRoutes).length>0,n=["AppRoutes","PageRoutes","LayoutRoutes","RedirectRoutes","RewriteRoutes","ParamMap"];a&&n.push("AppRouteHandlerRoutes");const r=`export type { ${n.join(", ")} }`;return`// This file is generated automatically by Next.js
// Do not edit this file manually

${o}

${t}

export type ParamsOf<Route extends Routes> = ParamMap[Route]

${s}

${r}

declare global {
  /**
   * Props for Next.js App Router page components
   * @example
   * \`\`\`tsx
   * export default function Page(props: PageProps<'/blog/[slug]'>) {
   *   const { slug } = await props.params
   *   return <div>Blog post: {slug}</div>
   * }
   * \`\`\`
   */
  interface PageProps<AppRoute extends AppRoutes> {
    params: Promise<ParamMap[AppRoute]>
    searchParams: Promise<Record<string, string | string[] | undefined>>
  }

  /**
   * Props for Next.js App Router layout components
   * @example
   * \`\`\`tsx
   * export default function Layout(props: LayoutProps<'/dashboard'>) {
   *   return <div>{props.children}</div>
   * }
   * \`\`\`
   */
  type LayoutProps<LayoutRoute extends LayoutRoutes> = {
    params: Promise<ParamMap[LayoutRoute]>
    children: React.ReactNode
  } & {
    [K in LayoutSlotMap[LayoutRoute]]: React.ReactNode
  }${a?`

  /**
   * Context for Next.js App Router route handlers
   * @example
   * \`\`\`tsx
   * export async function GET(request: NextRequest, context: RouteContext<'/api/users/[id]'>) {
   *   const { id } = await context.params
   *   return Response.json({ id })
   * }
   * \`\`\`
   */
  interface RouteContext<AppRouteHandlerRoute extends AppRouteHandlerRoutes> {
    params: Promise<ParamMap[AppRouteHandlerRoute]>
  }`:""}
}
`}function generateRouteTypesFileStrict(e){const o=generateRouteTypes(e),t=generateParamTypes(e),s=generateLayoutSlotMap(e),a=Object.keys(e.appRouteHandlerRoutes).length>0,n=["AppRoutes","PageProps","PageRoutes","LayoutRoutes","RedirectRoutes","RewriteRoutes","ParamMap"];a&&n.push("AppRouteHandlerRoutes");const r=`export type { ${n.join(", ")} }`;return`// This file is generated automatically by Next.js with experimental.strictRouteTypes
// Do not edit this file manually

${o}

${t}

export type ParamsOf<Route extends Routes> = ParamMap[Route]

${s}

${r}

declare global {
  /**
   * Props for Next.js App Router page components
   * @example
   * \`\`\`tsx
   * export default function Page(props: PageProps<'/blog/[slug]'>) {
   *   const { slug } = await props.params
   *   return <div>Blog post: {slug}</div>
   * }
   * \`\`\`
   */
  interface PageProps<AppRoute extends AppRoutes> {
    params: Promise<ParamMap[AppRoute]>
    searchParams: Promise<Record<string, string | string[] | undefined>>
  }

  /**
   * Props for Next.js App Router layout components
   * @example
   * \`\`\`tsx
   * export default function Layout(props: LayoutProps<'/dashboard'>) {
   *   return <div>{props.children}</div>
   * }
   * \`\`\`
   */
  type LayoutProps<LayoutRoute extends LayoutRoutes> = {
    params: Promise<ParamMap[LayoutRoute]>
    children: React.ReactNode
  } & {
    [K in LayoutSlotMap[LayoutRoute]]: React.ReactNode
  }${a?`

  /**
   * Context for Next.js App Router route handlers
   * @example
   * \`\`\`tsx
   * export async function GET(request: NextRequest, context: RouteContext<'/api/users/[id]'>) {
   *   const { id } = await context.params
   *   return Response.json({ id })
   * }
   * \`\`\`
   */
  interface RouteContext<AppRouteHandlerRoute extends AppRouteHandlerRoutes> {
    params: Promise<ParamMap[AppRouteHandlerRoute]>
  }`:""}
}
`}
